enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,441 @@
+between how Ingresses are supported by cert-manager via `.spec.tls.secretName` and when using
+Routes, the user would have to provide the Certificate CR reference.
+
+This also adds coupling with external third-party API(s) and limits future enhancement to be
+a more generic solution for integrating with third-party certificate management systems.
+
+##### Alternative to new status condition types
+
+An alternative to introducing 3 new condition types, would be to use `metadata.managedFields`
+on Route to denote to the user which fields are managed by the secret-injector controller and
+server-side apply would be used by the controller during updates to ensure that `.spec.tls.certificate`
+and `.spec.tls.secret` are always managed by the controller as long as the annotation is present.
+
+This although helps in avoiding addition of new condition types will result in a few inconsistencies,
+
+- When transitioning from using managed certificates to manually managing custom certificates,
+  the `.managedFields` will still depict `.spec.tls` as managed by the controller since the update
+  operation done by the user is only on the `.metadata.annotations`.
+- Also since `oc` defaults to `--server-side=false`, mixing CSA and SSA results in ambiguous `.managedFields`.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The new controller (secret-injector) will be responsible for watching routes and processing those
+that have the valid annotation for `tls-secret-name`. This controller will also update
+`.status` of the route to contain the latest information on the certificate in use.
+
+Since the secrets are the primary resource for the secret-injector controller, the new controller
+will also set up watches and re-sync routes in the same namespace as that of the secret (similar
+to ingress-to-route controller).
+
+The new controller will watch all routes and process all routes in order to update
+`.status.certificate`. The ones with the annotation will have additional
+logic to update `.spec.tls.certificate` and `.spec.tls.key` based on the following pre-conditions,
+
+- The secret-reference annotation is present.
+- Secret mentioned in the annotation exists and is valid (contains the required fields mentioned [here](#api-extensions))
+
+In scenarios where the user deletes the annotation, it is the controller's
+responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e.
+the latest certificate status condition has to be `ManagedCertificate` and since the required annotation is
+not present this results in the `.spec.tls.certificate` and `.spec.tls.key` being cleared
+out and the `.status` updated to `DefaultCertificate`.
+
+If the user deletes the secret associated with the Route without deleting the
+`route.openshift.io/tls-secret-name` annotation, the secret-injector controller will
+clear out `.spec.tls.certificate` and `.spec.tls.key` since the following pre-conditions
+are met,
```

Comment:
You could call it an error condition.  The user specified incomplete configuration (dangling cert reference), so the controller makes its best effort (use the default cert) and notifies the user of the issue via an event and status condition.  How is that opaque?  It's consistent with the behavior of the ingress-to-route controller (if you delete the secret that an ingress references, the ingress-to-route controller deletes any associated routes).

Reply:
(trying to remember what I was thinking 2 weeks ago...)

In terms of the route itself, what would a user see if they tried to access the endpoint? Possibly a certificate error? Possibly not? 

I think I was going down the line of, if the endpoint suddenly failed spectacularly, it would be obvious there was an error in the route, but that's probably not a great UX for something like a Route API, so your current behaviour from ingress-to-route is more pleasant in that regard